File: clients/python/harness/reporter.py
"""Writes reports/results.json and reports/summary.md from TestResults."""

from __future__ import annotations

import json
import os
from collections import Counter, defaultdict

from .result import FAIL, PASS, SKIP, TestResult


class Reporter:
    def __init__(self, results: list[TestResult], out_dir: str, engine: str):
        self.results = results
        self.out_dir = out_dir
        self.engine = engine

    def write(self) -> dict:
        os.makedirs(self.out_dir, exist_ok=True)
        total = len(self.results)
        npass = sum(1 for r in self.results if r.status == PASS)
        nfail = sum(1 for r in self.results if r.status == FAIL)
        nskip = sum(1 for r in self.results if r.status == SKIP)

        payload = {
            "engine": self.engine,
            "summary": {"total": total, "pass": npass, "fail": nfail, "skip": nskip},
            "results": [r.to_dict() for r in self.results],
        }
        with open(os.path.join(self.out_dir, "results.json"), "w", encoding="utf-8") as f:
            json.dump(payload, f, indent=2, ensure_ascii=False)

        with open(os.path.join(self.out_dir, "summary.md"), "w", encoding="utf-8") as f:
            f.write(self._markdown(total, npass, nfail, nskip))

        return {"total": total, "pass": npass, "fail": nfail, "skip": nskip}

    def _markdown(self, total, npass, nfail, nskip) -> str:
        out: list[str] = []
        out.append("# Python ORM ⇆ MatrixOne Compatibility Summary")
        out.append("")
        out.append(f"- **Engine:** `{self.engine}`")
        out.append(f"- **Scenarios:** {total} total — "
                   f"**{npass} pass**, **{nfail} fail**, **{nskip} skip**")
        denom = max(1, total - nskip)
        out.append(f"- **Pass rate (excl. skip):** {npass / denom * 100:.1f}%")
        out.append("")

        # By framework
        out.append("## Pass/fail by framework")
        out.append("")
        out.append("| Framework | Total | Pass | Fail | Skip | Pass % |")
        out.append("|---|--:|--:|--:|--:|--:|")
        by_fw: dict[str, list[TestResult]] = defaultdict(list)
        for r in self.results:
            by_fw[r.framework].append(r)
        for fw in sorted(by_fw):
            rs = by_fw[fw]
            t = len(rs)
            p = sum(1 for r in rs if r.status == PASS)
            fl = sum(1 for r in rs if r.status == FAIL)
            sk = sum(1 for r in rs if r.status == SKIP)
            pr = p / max(1, t - sk) * 100
            out.append(f"| {fw} | {t} | {p} | {fl} | {sk} | {pr:.1f}% |")
        out.append("")

        # By category
        out.append("## Pass/fail by category")
        out.append("")
        out.append("| Category | Total | Pass | Fail | Skip |")
        out.append("|---|--:|--:|--:|--:|")
        by_cat: dict[str, list[TestResult]] = defaultdict(list)
        for r in self.results:
            by_cat[r.category].append(r)
        for cat in sorted(by_cat):
            rs = by_cat[cat]
            t = len(rs)
            p = sum(1 for r in rs if r.status == PASS)
            fl = sum(1 for r in rs if r.status == FAIL)
            sk = sum(1 for r in rs if r.status == SKIP)
            out.append(f"| {cat} | {t} | {p} | {fl} | {sk} |")
        out.append("")

        # Failures by error signature
        out.append("## Failures by error signature")
        out.append("")
        code_counts: Counter = Counter()
        code_sample: dict[str, str] = {}
        for r in self.results:
            if r.status == FAIL:
                code = r.error_code or "ERROR"
                code_counts[code] += 1
                code_sample.setdefault(code, (r.error_message or "")[:200])
        if not code_counts:
            out.append("_No failures recorded._")
        else:
            out.append("| Error code | Count | Sample message |")
            out.append("|---|--:|---|")
            for code, cnt in code_counts.most_common():
                sample = _md_escape(code_sample.get(code, ""))
                out.append(f"| `{code}` | {cnt} | {sample} |")
        out.append("")

        # Harness-bug callout (should be zero)
        hb = [r for r in self.results if r.status == FAIL and r.error_code == "HARNESS-BUG"]
        out.append("## Harness-bug failures (must be zero)")
        out.append("")
        if not hb:
            out.append("**0 harness-bug failures.** All FAILs are DB-driven incompatibilities.")
        else:
            out.append(f"**{len(hb)} harness-bug failures — these are bugs in the suite:**")
            out.append("")
            for r in hb[:50]:
                out.append(f"- #{r.id} {r.framework}/{r.category}/{r.name}: "
                           f"{_md_escape((r.error_message or '')[:160])}")
        out.append("")

        # Top distinct failure messages
        out.append("## Top distinct failure messages")
        out.append("")
        msg_counts: Counter = Counter()
        for r in self.results:
            if r.status == FAIL:
                msg_counts[_norm_msg(r.error_message or "")] += 1
        out.append("| Count | Message |")
        out.append("|--:|---|")
        for msg, cnt in msg_counts.most_common(30):
            out.append(f"| {cnt} | {_md_escape(msg[:180])} |")
        out.append("")

        return "\n".join(out) + "\n"


def _norm_msg(msg: str) -> str:
    import re
    msg = re.sub(r"\s+", " ", msg).strip()
    # Collapse table/column suffixes so similar messages group together.
    msg = re.sub(r"__?[a-z0-9]{6,}\b", "<id>", msg)
    msg = re.sub(r"\b\d{4,}\b", "<n>", msg)
    return msg


def _md_escape(s: str) -> str:
    return s.replace("|", "\\|").replace("\n", " ")

File: clients/java/src/main/java/mo/harness/Reporter.java
package mo.harness;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/** Writes reports/results.json and reports/summary.md. */
public final class Reporter {

    private final String engine;
    private final List<Result> results;

    public Reporter(String engine, List<Result> results) {
        this.engine = engine;
        this.results = results;
    }

    public void write(Path dir) throws IOException {
        Files.createDirectories(dir);
        writeJson(dir.resolve("results.json"));
        writeMarkdown(dir.resolve("summary.md"));
    }

    private void writeJson(Path file) throws IOException {
        Map<String, Object> root = new LinkedHashMap<>();
        root.put("engine", engine);

        Map<String, Object> summary = new LinkedHashMap<>();
        int pass = 0, fail = 0, skip = 0;
        for (Result r : results) {
            switch (r.status) {
                case "PASS" -> pass++;
                case "FAIL" -> fail++;
                case "SKIP" -> skip++;
            }
        }
        summary.put("total", results.size());
        summary.put("pass", pass);
        summary.put("fail", fail);
        summary.put("skip", skip);
        root.put("summary", summary);

        List<Map<String, Object>> arr = new ArrayList<>();
        for (Result r : results) {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("framework", r.framework);
            m.put("category", r.category);
            m.put("name", r.name);
            m.put("status", r.status);
            m.put("duration_ms", r.durationMs);
            m.put("error_code", r.errorCode);
            m.put("error_message", r.errorMessage);
            arr.add(m);
        }
        root.put("results", arr);

        Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
        Files.writeString(file, gson.toJson(root), StandardCharsets.UTF_8);
    }

    private void writeMarkdown(Path file) throws IOException {
        StringBuilder sb = new StringBuilder();
        int pass = 0, fail = 0, skip = 0;
        for (Result r : results) {
            switch (r.status) {
                case "PASS" -> pass++;
                case "FAIL" -> fail++;
                case "SKIP" -> skip++;
            }
        }
        int total = results.size();
        sb.append("# MatrixOne Java ORM Compatibility — Summary\n\n");
        sb.append("- **Engine:** ").append(engine).append("\n");
        sb.append("- **Total scenarios run:** ").append(total).append("\n");
        sb.append("- **Pass:** ").append(pass)
                .append(" (").append(pct(pass, total)).append(")\n");
        sb.append("- **Fail:** ").append(fail)
                .append(" (").append(pct(fail, total)).append(")\n");
        sb.append("- **Skip:** ").append(skip)
                .append(" (").append(pct(skip, total)).append(")\n\n");

        // Pass/fail by framework
        sb.append("## Pass / fail by framework\n\n");
        sb.append("| Framework | Pass | Fail | Skip | Pass % |\n");
        sb.append("|---|--:|--:|--:|--:|\n");
        Map<String, int[]> byFw = new TreeMap<>();
        for (Result r : results) {
            int[] c = byFw.computeIfAbsent(r.framework, k -> new int[3]);
            switch (r.status) {
                case "PASS" -> c[0]++;
                case "FAIL" -> c[1]++;
                case "SKIP" -> c[2]++;
            }
        }
        for (var e : byFw.entrySet()) {
            int[] c = e.getValue();
            int t = c[0] + c[1] + c[2];
            sb.append("| ").append(e.getKey())
                    .append(" | ").append(c[0])
                    .append(" | ").append(c[1])
                    .append(" | ").append(c[2])
                    .append(" | ").append(pct(c[0], t)).append(" |\n");
        }
        sb.append("\n");

        // Pass/fail by category
        sb.append("## Pass / fail by category\n\n");
        sb.append("| Category | Pass | Fail | Skip |\n");
        sb.append("|---|--:|--:|--:|\n");
        Map<String, int[]> byCat = new TreeMap<>();
        for (Result r : results) {
            int[] c = byCat.computeIfAbsent(r.category, k -> new int[3]);
            switch (r.status) {
                case "PASS" -> c[0]++;
                case "FAIL" -> c[1]++;
                case "SKIP" -> c[2]++;
            }
        }
        for (var e : byCat.entrySet()) {
            int[] c = e.getValue();
            sb.append("| ").append(e.getKey())
                    .append(" | ").append(c[0])
                    .append(" | ").append(c[1])
                    .append(" | ").append(c[2]).append(" |\n");
        }
        sb.append("\n");

        // Failures by error signature
        sb.append("## Failures by error signature (error_code)\n\n");
        sb.append("| Error code | Count | Meaning |\n");
        sb.append("|---|--:|---|\n");
        Map<String, Integer> bySig = new LinkedHashMap<>();
        for (Result r : results) {
            if (!"FAIL".equals(r.status)) continue;
            String key = r.errorCode == null ? "(none)" : r.errorCode;
            bySig.merge(key, 1, Integer::sum);
        }
        bySig.entrySet().stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                .forEach(e -> sb.append("| ").append(e.getKey())
                        .append(" | ").append(e.getValue())
                        .append(" | ").append(meaning(e.getKey())).append(" |\n"));
        sb.append("\n");

        // Top distinct failure messages
        sb.append("## Top distinct failure messages\n\n");
        Map<String, Integer> byMsg = new LinkedHashMap<>();
        Map<String, String> sampleCode = new LinkedHashMap<>();
        for (Result r : results) {
            if (!"FAIL".equals(r.status)) continue;
            String key = normalizeMsg(r.errorMessage);
            byMsg.merge(key, 1, Integer::sum);
            sampleCode.putIfAbsent(key, r.errorCode);
        }
        sb.append("| Count | Code | Message (normalized) |\n");
        sb.append("|--:|---|---|\n");
        byMsg.entrySet().stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                .limit(40)
                .forEach(e -> sb.append("| ").append(e.getValue())
                        .append(" | ").append(sampleCode.get(e.getKey()))
                        .append(" | ").append(escapePipe(e.getKey())).append(" |\n"));
        sb.append("\n");

        // Behavior mismatches (data-integrity findings) listed explicitly
        sb.append("## BEHAVIOR mismatches (ran OK but differs from MySQL)\n\n");
        sb.append("| Framework | Name | Detail |\n");
        sb.append("|---|---|---|\n");
        List<Result> behaviors = new ArrayList<>();
        for (Result r : results) {
            if ("FAIL".equals(r.status) && "BEHAVIOR".equals(r.errorCode)) behaviors.add(r);
        }
        behaviors.sort(Comparator.comparing(r -> r.name));
        for (Result r : behaviors) {
            sb.append("| ").append(r.framework)
                    .append(" | ").append(escapePipe(r.name))
                    .append(" | ").append(escapePipe(r.errorMessage)).append(" |\n");
        }
        sb.append("\n");

        Files.writeString(file, sb.toString(), StandardCharsets.UTF_8);
    }

    private static String normalizeMsg(String msg) {
        if (msg == null) return "(no message)";
        // Strip variable identifiers (table names with random suffixes, numbers).
        String n = msg.replaceAll("\\b\\d{3,}\\b", "#")
                .replaceAll("_[0-9a-f]{6,}", "_X")
                .replaceAll("'[^']*'", "'_'");
        return ErrorClassifier.truncate(n);
    }

    private static String meaning(String code) {
        return switch (code) {
            case "20105" -> "function / operator not implemented";
            case "1064" -> "SQL parser / syntax not supported";
            case "BEHAVIOR" -> "ran OK but result differs from MySQL semantics";
            case "20101" -> "internal 'not implemented yet' (e.g. ROLLBACK TO SAVEPOINT)";
            case "20203" -> "stricter argument / type validation than MySQL";
            case "1690" -> "out-of-range / overflow";
            case "1062" -> "duplicate entry for key";
            case "SKIP" -> "skipped (precondition)";
            default -> "";
        };
    }

    private static String escapePipe(String s) {
        if (s == null) return "";
        return s.replace("|", "\\|").replace("\n", " ");
    }

    private static String pct(int n, int total) {
        if (total == 0) return "0.0%";
        return String.format("%.1f%%", 100.0 * n / total);
    }
}
